### PR TITLE
Adding Subscription Metrics to Metric Recorder and Onboarding M365 to Auth Metrics from Metrics Recorder

### PR DESCRIPTION
**Pull request by chrisale000**
### Description
This PR introduces comprehensive metrics tracking for Microsoft Office 365 API operations in the Data Prepper M365 source plugin by extending the VendorAPIMetricsRecorder to support both start subscription APIs and authentication APIs, maintaining centralized vendor API metrics while adding M365-specific operation tracking.

#### New Metrics in VendorAPIMetricsRecorder
- Start Subscription Metrics: Extended VendorAPIMetricsRecorder to support Office 365 start subscription operations
    - Tracks start subscription operations (success/failure/latency/call counts)
    - Provides multiple overloaded methods for latency recording (Supplier, Runnable, Duration)
    - Maintains centralized vendor API metrics while supporting M365-specific operations

#### Modified Components
- Office365RestClient: Updated to use VendorAPIMetricsRecorder for subscription metrics
   - Integrated VendorAPIMetricsRecorder for start subscription operations
   - Wrapped start subscription logic with latency tracking
   - Removed generic apiCallsCounter in favor of specific start subscription metrics
- Office365Configuration (formerly Office365RestClientConfiguration): Updated dependency injection
   - Updated Office365RestClient bean to inject VendorAPIMetricsRecorder for subscription metrics
   - Added VendorAPIMetricsRecorder bean injection for Office365AuthenticationProvider
   - Renamed from Office365RestClientConfiguration to reflect broader configuration scope

#### Test Coverage
- VendorAPIMetricsRecorderTest: Updated/added unit tests
    - Tests subscription metric recording methods
    - Validates success/failure tracking for start subscription operations
    - Tests latency recording with multiple overloads
    - Includes edge cases and error scenarios
- Office365RestClientTest: Updated integration tests
    - Added tests for metrics recording during start subscription operations
    - Validates failure metric recording
    - Tests AF20024 error handling with metrics
- Office365AuthenticationProviderTest: Updated unit tests
    - Added tests for VendorAPIMetricsRecorder injection
    - Validates authentication metrics recording
    - Tests success and failure scenarios for authentication metrics

#### Metrics Added
- Start Subscription Metrics(via VendorAPIMetricsRecorder):
   - startSubscriptionRequestsSuccess: Count of successful start subscription operations
   - startSubscriptionRequestsFailed: Count of failed start subscription operations
   - startSubscriptionRequestLatency: Timer for start subscription operation duration
   - startSubscriptionApiCalls: Count of individual API calls during start subscription operations
- Authentication Metrics (via VendorAPIMetricsRecorder):
   - authenticationRequestsSuccess: Count of successful authentication requests
   - authenticationRequestsFailed: Count of failed authentication requests
   - authenticationRequestLatency: Timer for authentication request duration
### Testing: 
- Spun up local M365 pipeline and confirmed that all metrics (subscription and authentication) were generated successfully.
 
### Check List
- [X] New functionality includes testing.
- [X] New functionality has a documentation issue. Please link to it in this PR.
  - [X] New functionality has javadoc added
- [X] Commits are signed with a real name per the DCO

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/data-prepper/blob/main/CONTRIBUTING.md).

### CODE REVIEW

File: data-prepper-plugins/saas-source-plugins/microsoft-office365-source/src/main/java/org/opensearch/dataprepper/plugins/source/microsoft_office365/Office365RestClient.java [hunk 1]
```diff
@@ -69,59 +66,66 @@ public Office365RestClient(final Office365AuthenticationInterface authConfig,
...
+//                        HttpMethod.GET,
+//                        new HttpEntity<>(headers),
+//                        String.class
+//                );
+//                log.debug("Current subscriptions: {}", listResponse.getBody());
 
-            // Start subscriptions for each content type
-            headers.setContentLength(0);
+                // Start subscriptions for each content type
+                headers.setContentLength(0);
```
Comment (vecheka) [inline data-prepper-plugins/saas-source-plugins/microsoft-office365-source/src/main/java/org/opensearch/dataprepper/plugins/source/microsoft_office365/Office365RestClient.java:86]: nit: indentation.

File: data-prepper-plugins/saas-source-plugins/microsoft-office365-source/src/main/java/org/opensearch/dataprepper/plugins/source/microsoft_office365/metrics/Office365MetricsRecorder.java [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+ * - StartSubscription success/failure rates: Track successful and failed subscription operations
+ * - StartSubscription call counts: Track individual API calls made for subscription management
+ * 
+ * This is separate from the shared VendorAPIMetricsRecorder to isolate Office 365-specific
+ * API metrics that are unique to the Office 365 Management API within the M365 module.
+ * 
+ * NOTE: Any new Office 365 specific metrics should be implemented here to maintain proper
+ * separation of vendor-specific functionality and keep Office 365 metrics centralized.
+ */
+public class Office365MetricsRecorder {
```
Comment (wjyao0316) [inline data-prepper-plugins/saas-source-plugins/microsoft-office365-source/src/main/java/org/opensearch/dataprepper/plugins/source/microsoft_office365/metrics/Office365MetricsRecorder.java]: It would be hard to maintain metric allowlist if we start to introduce per service metric. 

We should try to make this metric generic so can be extended for similar cases in the future. 

I would suggest add StartSubscription recorder to the vendorAPIMetric instead of having its own one.

File: data-prepper-plugins/saas-source-plugins/microsoft-office365-source/src/main/java/org/opensearch/dataprepper/plugins/source/microsoft_office365/Office365RestClient.java (hunk 1 above)
Comment (chrisale000) [reply data-prepper-plugins/saas-source-plugins/microsoft-office365-source/src/main/java/org/opensearch/dataprepper/plugins/source/microsoft_office365/Office365RestClient.java:86]: The indentation was previously off, this fixes the incorrect indentation.

Comment (graytaylor0) [reply data-prepper-plugins/saas-source-plugins/microsoft-office365-source/src/main/java/org/opensearch/dataprepper/plugins/source/microsoft_office365/Office365RestClient.java:86]: I think in the future you can run this command to fix indentation

```
./gradlew :data-prepper-plugins:saas-source-plugins:microsoft-office365-source:spotlessApply
```

Comment (chrisale000) [reply data-prepper-plugins/saas-source-plugins/microsoft-office365-source/src/main/java/org/opensearch/dataprepper/plugins/source/microsoft_office365/Office365RestClient.java:86]: Awesome, thank you for this command. Just ran it and looks like no changes were made so the new indentation is correct!
